physical/redis.go

```diff
@@ -0,0 +1,370 @@
+		if slashPos == -1 {
+			// We have an ordinary file, so return it directly.
+			results = append(results, match)
+		} else {
+			// Keep the trailing slash.
+			dir := match[0 : slashPos+1]
+
+			// Have we seen this directory already?  If not,
+			// return it and mark it as seen.
+			if _, present := dirs[dir]; !present {
+				results = append(results, dir)
+				dirs[dir] = true
+			}
+		}
+	}
+
+	return results, err
+}
+
+func (c *RedisBackend) LockWith(key, value string) (Lock, error) {
+	return &RedisLock{
+		key:      fmt.Sprintf("%s/_lock/%s", c.path, key),
+		value:    value,
+		pool:     c.pool,
+		logger:   c.logger,
+		lockConf: c.lockConf,
+	}, nil
+}
+
+// A lock implemented using Redis.  This is based on the single-Redis
+// example code used to explain Redlock at http://redis.io/topics/distlock
+// but it might be useful to upgrade this to a full Redlock implementation.
+type RedisLock struct {
+	key      string
+	value    string
+	pool     *redis.Pool
+	logger   *log.Logger
+	lockConf *RedisLockConfig
+}
+
+// Acquire the lock.  To interrupt lock acquistion, close stopCh.  The
+// returned channel will be closed if the lock is lost.
+func (c *RedisLock) Lock(stopCh <-chan struct{}) (<-chan struct{}, error) {
+	// Round up leaderTTL to the nearest second and convert to a
+	// string.  We do this because EXPIRE only has 1-second accuracy.
+	ttlSeconds := float64(c.lockConf.leaderTTL) / float64(time.Second);
+	ttlStr := strconv.Itoa(int(math.Ceil(ttlSeconds)));
+
+	// Loop until we get a non-nil reply from Redis indicating a
+	// successful operation.
+	var reply interface{}
+	for reply == nil {
+		conn := c.pool.Get()
+		defer conn.Close()
+
+		// Attempt to set our key.  "NX" means to only set the key
+		// if it does not exist, and "EX" specifies a timeout in
+		// seconds.
+		reply, err := conn.Do("SET", c.key, c.value, "NX", "EX", ttlStr)
+		if err != nil {
+			// We got an error communicating with Redis, so
+			// fail outright.
+			return nil, err
+		}
+		if reply != nil {
+			// We got back a non-nil response, which means we
+			// set the key and can stop looping.
+			break
+		}
+
+		// Wait a while before retrying.
+		select {
+		case <-stopCh:
+			// Lock acquisition was cancelled by our caller.
+			return nil, nil
+		case <-time.After(c.lockConf.leaderLockRetryInterval):
+			// Timeout fired, so here we go again.
+		}
+	}
+
+	// Set up a background listener to renew our lock and notice if it
+	// goes away.
+	leaderCh := make(chan struct{})
+	go func() {
+		conn := c.pool.Get()
+		defer conn.Close()
+
+		// Create our TTL-bumping script and try to load
+		// it. Loading is optional because of how Redigo is
+		// implemented.
+		bumpTtlScript := redis.NewScript(1, `
+if redis.call("GET", KEYS[1]) == ARGV[1] then
+    return redis.call("EXPIRE", KEYS[1], ARGV[2])
+else
+    return 0
+end
+`)
+		_ = bumpTtlScript.Load(conn)
+
+		// Renew our TTL periodically.
+		for {
+			time.Sleep(c.lockConf.leaderTTLRenewInterval)
```

Comment:
Go uses `time.Duration` to represent all durations in a standardized fashion. We follow this convention.

AFAIK, `time.Duration` is something like `int64` nanoseconds. But I'm supposed to pretend it's a completely opaque duration type.